Rubberduck.CodeAnalysis/CodePathAnalysis/Walker.cs

```diff
@@ -0,0 +1,88 @@
+﻿using Antlr4.Runtime.Tree;
+using Rubberduck.Inspections.CodePathAnalysis.Nodes;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Symbols;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Rubberduck.Inspections.CodePathAnalysis
+{
+    public class Walker
+    {
+        public INode GenerateTree(IParseTree tree, Declaration declaration)
+        {
+            INode node = default;
+            switch (tree)
+            {
+                case VBAParser.ForNextStmtContext _:
+                case VBAParser.ForEachStmtContext _:
+                case VBAParser.WhileWendStmtContext _:
+                case VBAParser.DoLoopStmtContext _:
+                    node = new LoopNode();
+                    break;
+                case VBAParser.IfStmtContext _:
+                case VBAParser.ElseBlockContext _:
+                case VBAParser.ElseIfBlockContext _:
+                case VBAParser.SingleLineIfStmtContext _:
+                case VBAParser.SingleLineElseClauseContext _:
+                case VBAParser.CaseClauseContext _:
+                case VBAParser.CaseElseClauseContext _:
+                    node = new BranchNode();
+                    break;
+                case VBAParser.BlockContext _:
+                    node = new BlockNode();
+                    break;
+            }
+
+            if (declaration.Context == tree)
+            {
+                node = new DeclarationNode
+                {
+                    Declaration = declaration
+                };
+            }
+
+            var reference = declaration.References.SingleOrDefault(w => w.Context == tree);
+            if (reference != null)
+            {
+                if (reference.IsAssignment)
+                {
+                    node = new AssignmentNode
+                    {
+                        Reference = reference
+                    };
+                }
+                else
+                {
+                    node = new ReferenceNode
+                    {
+                        Reference = reference
+                    };
+                }
+            }
+
+            if (node == default)
+            {
+                node = new GenericNode();
+            }
+
+            var children = new List<INode>();
+            for (var i = 0; i < tree.ChildCount; i++)
+            {
+                var nextChild = GenerateTree(tree.GetChild(i), declaration);
+                nextChild.SortOrder = i;
+                nextChild.Parent = node;
+
+                if (nextChild.Children.Any() || nextChild.GetType() != typeof(GenericNode))
```

Comment:
GenericNode acts as a substitute for `default` here. I'm not sure I like that...

Reply:
@Vogel612 except `default` would be `null`, but `GenericNode` is still a node that can have interesting child nodes, or am I misreading this?

Reply:
@retailcoder GenericNode's children are fixed to an empty list. GenericNode also is the fallback when we don't assign any other node type. This makes sense on one hand, on the other hand they are explicitly excluded from being added to the Children collection when the tree is generated...

Reply:
@Vogel612 No, that's not quite right. Mat's answer is correct that it can have interesting child nodes. This is just here because it generates a `GenericNode` for every single node in the tree, and 90% of the nodes weren't considered special *yet*. This meant that to check the tree in the debugger, I had to run down a million rabbit holes, and everywhere I tried to analyze it, I'd have to spin through a million `GenericNode` values.

This block only excludes the node if A) it is a generic node and B) there are no children. It basically prunes any uninteresting branches off the tree and leaves only the relevant ones for the analysis.